### PR TITLE
newdevice: APC Environmental monitoring units #5140

### PR DESCRIPTION
**Pull request by laf**
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/)
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

Fixes: #5140

Few things.

1. These devices allow users to specify the data is in Fahrenheit rather than Celsius so I've added a function to convert.

2. The includes/polling/sensors/X/ folders were named wrong so have been moved. They don't follow the plural like disco but use the name registered in discover_sensor() which is the singular!

3. Added a pre-cache for sensor polling.

### CODE REVIEW

File: includes/discovery/sensors/humidity/aos-emu2.inc.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+if ($device['os'] === 'aos-emu2') {
+    foreach ($emu2_temp as $id => $temp) {
+        if (isset($temp['emsProbeStatusProbeHumidity']) && $temp['emsProbeStatusProbeHumidity'] > 0) {
+            $index           = $temp['emsProbeStatusProbeIndex'];
+            $oid             = '.1.3.6.1.4.1.318.1.1.10.3.13.1.1.3.' . $index;
+            $descr           = $temp['emsProbeStatusProbeName'];
+            $low_limit       = $temp['emsProbeStatusProbeMinHumidityThresh'];
+            $low_warn_limit  = $temp['emsProbeStatusProbeLowHumidityThresh'];
+            $high_limit      = $temp['emsProbeStatusProbeMaxHumidityThresh'];
```
Comment (Rosiak) [inline includes/discovery/sensors/humidity/aos-emu2.inc.php]: The device I tested against doesn't send any info back for:
emsProbeStatusProbeMaxHumidityThresh
emsProbeStatusProbeMinHumidityThresh

File: includes/discovery/sensors/humidity/aos-emu2.inc.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * @link       http://librenms.org
+ * @copyright  2016 Neil Lathwood
+ * @author     Neil Lathwood <neil@lathwood.co.uk>
+ */
+
+if ($device['os'] === 'aos-emu2') {
+    foreach ($emu2_temp as $id => $temp) {
+        if (isset($temp['emsProbeStatusProbeHumidity']) && $temp['emsProbeStatusProbeHumidity'] > 0) {
+            $index           = $temp['emsProbeStatusProbeIndex'];
+            $oid             = '.1.3.6.1.4.1.318.1.1.10.3.13.1.1.3.' . $index;
```
Comment (Rosiak) [inline includes/discovery/sensors/humidity/aos-emu2.inc.php]: I belive this should be .1.3.6.1.4.1.318.1.1.10.3.13.1.1.6.

File: includes/discovery/sensors/temperatures/aos-emu2.inc.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * @author     Neil Lathwood <neil@lathwood.co.uk>
+ */
+
+if ($device['os'] === 'aos-emu2') {
+    foreach ($emu2_temp as $id => $temp) {
+        if (isset($temp['emsProbeStatusProbeTemperature']) && $temp['emsProbeStatusProbeTemperature'] > 0) {
+            $index           = $temp['emsProbeStatusProbeIndex'];
+            $oid             = '.1.3.6.1.4.1.318.1.1.10.3.13.1.1.3.' . $index;
+            $descr           = $temp['emsProbeStatusProbeName'];
+            $low_limit       = fahrenheit_to_celsius($emu2_temp_scale, $temp['emsProbeStatusProbeMinTempThresh']);
```
Comment (Rosiak) [inline includes/discovery/sensors/temperatures/aos-emu2.inc.php]: The device I tested against doesn't send any info back for:
emsProbeStatusProbeMaxTempThresh
emsProbeStatusProbeMinTempThresh

File: includes/discovery/sensors/humidity/aos-emu2.inc.php (hunk 2 above)
Comment (laf) [inline includes/discovery/sensors/humidity/aos-emu2.inc.php]: well spotted, updated now

File: includes/discovery/sensors/humidity/aos-emu2.inc.php (hunk 1 above)
Comment (laf) [inline includes/discovery/sensors/humidity/aos-emu2.inc.php]: does it not then just auto calculate?

**Comment by Rosiak** [comment]
Disco: http://pastebin.com/2hgQtvVA
Poll: http://pastebin.com/8QCDH7sX

**Comment by laf** [comment]
Forgot to push this last night. Added a function to set the default value to null based on the minimum we expect so if none of the thresholds are valid (i.e 0 according to your output) then we default to null so they are excluded.

The Fahrenheit was fixed in another commit but you were on the older one hence it didn't work for you.

**Comment by laf** [comment]
bump

**Comment by laf** [comment]
fixed rebase.

**Comment by Rosiak** [comment]
Last thing:
Device I have access to does not respond on emsProbeStatusProbeMaxHumidityThresh or  emsProbeStatusProbeMinHumidityThresh

**Comment by laf** [comment]
Thanks, I've moved set_null to inside discover_sensor now so any invalid high/low limits will be nulled out.

**Comment by Rosiak** [comment]
Still an issue, you are aware that it was Humidity that was wrong, right?
Temp polling/warnings is perfectly fine :)

**Comment by laf** [comment]
Yes but the call is now in discover_sensor() so cleans anything. Up what's actually wrong and where's the debug on the latest commit?

**Comment by laf** [comment]
Can we look at merging this. The issue for @Rosiak is not from this PR, defaults for humidity sensors with no thresholds are 70/70

**Pull request merged**